bindings/rust/bench/src/harness.rs

```diff
@@ -1,163 +1,356 @@
 
-    /// Checks if handshake is finished for both client and server
     fn handshake_completed(&self) -> bool;
 
-    /// Get negotiated cipher suite
     fn get_negotiated_cipher_suite(&self) -> CipherSuite;
 
-    /// Get whether or negotiated version is TLS1.3
     fn negotiated_tls13(&self) -> bool;
 
-    /// Send application data from connection in harness pair
-    fn send(&mut self, sender: Mode, data: &[u8]) -> Result<(), Box<dyn Error>>;
+    /// Send application data to ConnectedBuffer
+    fn send(&mut self, data: &[u8]) -> Result<(), Box<dyn Error>>;
+
+    /// Read application data from ConnectedBuffer
+    fn recv(&mut self, data: &mut [u8]) -> Result<(), Box<dyn Error>>;
+
+    /// Shrink buffers owned by the connection
+    fn shrink_connection_buffers(&mut self);
 
-    /// Receive application data sent to connection in harness pair
-    fn recv(&mut self, receiver: Mode, data: &mut [u8]) -> Result<(), Box<dyn Error>>;
+    /// Clear and shrink buffers used for IO with another connection
+    fn shrink_connected_buffer(&mut self);
 
-    /// Send data from client to server and then from server to client
-    fn round_trip_transfer(&mut self, data: &mut [u8]) -> Result<(), Box<dyn Error>> {
+    /// Get reference to internal connected buffer
+    fn connected_buffer(&self) -> &ConnectedBuffer;
+}
+
+pub struct TlsConnPair<C: TlsConnection, S: TlsConnection> {
+    client: C,
+    server: S,
+}
+
+impl<C: TlsConnection, S: TlsConnection> Default for TlsConnPair<C, S> {
+    fn default() -> Self {
+        Self::new(Default::default(), Default::default(), Default::default()).unwrap()
+    }
+}
+
+impl<C: TlsConnection, S: TlsConnection> TlsConnPair<C, S> {
+    /// Wrap two TlsConnections into a TlsConnPair
+    pub fn wrap(client: C, server: S) -> Self {
+        assert!(
+            client.connected_buffer() == &server.connected_buffer().clone_inverse(),
+            "connected buffers don't match"
+        );
+        Self { client, server }
+    }
+
+    /// Take back ownership of individual connections in the TlsConnPair
+    pub fn unwrap(self) -> (C, S) {
+        (self.client, self.server)
+    }
+
+    /// Initialize buffers, configs, and connections (pre-handshake)
+    pub fn new(
+        crypto_config: CryptoConfig,
+        handshake_type: HandshakeType,
+        connected_buffer: ConnectedBuffer,
+    ) -> Result<Self, Box<dyn Error>> {
+        Ok(Self {
+            client: C::new(
+                Mode::Client,
+                crypto_config,
+                handshake_type,
+                connected_buffer.clone_inverse(),
+            )?,
+            server: S::new(
+                Mode::Server,
+                crypto_config,
+                handshake_type,
+                connected_buffer,
+            )?,
+        })
+    }
+
+    /// Run handshake on connections
+    pub fn handshake(&mut self) -> Result<(), Box<dyn Error>> {
+        for _ in 0..2 {
```

Comment:
I wonder whether it's worth adding a comment about why we still need 2 rtt's for TLS 1.3 (waiting for the server finished message)

Reply:
Yeah, I think this is a good place to call it out. I'll add it.